ssg/templates.py

```diff
@@ -112,198 +122,113 @@ def __init__(
         """
-        if rule.template is None:
-            return None
+        if lang not in self.templates[template_name].langs:
+            raise ValueError("Language {0} is not available for template {1}."
+                             .format(lang.name, template_name))
 
-        rule_langs = set(self.get_langs_to_generate(rule))
-        template_name = self.get_template_name(rule.template, rule.id_)
-        template_langs = set(templates[template_name].langs)
-        return rule_langs.intersection(template_langs)
+        template_file_name = lang.name + ".template"
+        template_file_path = os.path.join(self.templates_dir, template_name, template_file_name)
+        template_parameters = self.templates[template_name].preprocess(template_vars, lang.name)
+        env_yaml = self.env_yaml.copy()
+        env_yaml.update(local_env_yaml)
+        jinja_dict = ssg.utils.merge_dicts(env_yaml, template_parameters)
+        return ssg.jinja.process_file_with_macros(template_file_path, jinja_dict)
 
-    def process_product_vars(self, all_variables):
+    def get_templatable_lang_contents(self, templatable, language):
         """
-        Given a dictionary with the format key[@<product>]=value, filter out
-        and only take keys that apply to this product (unqualified or qualified
-        to exactly this product). Returns a new dict.
+        For the specified Templatable, build and return only the specified language content.
         """
-        processed = dict(filter(lambda item: '@' not in item[0], all_variables.items()))
-        suffix = '@' + self.env_yaml['product']
-        for variable in filter(lambda key: key.endswith(suffix), all_variables):
-            new_variable = variable[:-len(suffix)]
-            value = all_variables[variable]
-            processed[new_variable] = value
+        template_name = templatable.get_template_name()
+        template_vars = templatable.get_template_vars(self.env_yaml)
 
-        return processed
+        # Checks and remediations are processed with a custom YAML dict
+        local_env_yaml = templatable.get_template_context(self.env_yaml)
+        try:
+            return self.process_template_lang_file(template_name, template_vars, language, local_env_yaml)
+        except Exception as e:
+            raise RuntimeError("Unable to generate {0} template language for Templatable {1}: {2}"
+                               .format(language.name, templatable, e))
+
+    def write_templatable_lang_contents(self, filled_template, lang, templatable):
+        output_file_name = templatable.id_ + lang.file_extension
+        output_filepath = os.path.join(self.output_dirs[lang.name], output_file_name)
+        with open(output_filepath, "w") as f:
+            f.write(filled_template)
 
-    def build_rule(self, rule_id, rule_title, template, langs_to_generate, identifiers,
-                   platforms=None):
+    def build_templatable_lang(self, templatable, lang):
```

Comment:
I'd prefer `build_lang_for_templatable`.

Reply:
Sure, can do that.